### PR TITLE
Work towards Menu as well as Desktop links

### PR DESCRIPTION
**Pull request by maphew**
_Nutshell: I prefer Menu links over Desktop links as too many icons on the desktop distract me. So the `menu-links` branch is about adding the option of creating these._ 

This isn't ready to merge. It's okay if it were merged as it doesn't hurt or change any existing behaviour. I created the PR because it's the easiest way I know to see the differences and talk about where I'm aiming.

New things: `get_menu_folder()` and `get_desktop_folder()` are succinct means to get the full path to these folders on Windows regardless of locale (language), Windows version,and  [user moved Desktop somewhere else](https://winaero.com/blog/move-desktop-folder-windows-10/).
```
import win32com.client
shellapp = win32com.client.Dispatch("Shell.Application")
def get_menu_folder():
    '''Return user Start Menu folder'''
    return shellapp.namespace(11).self.path
def get_desktop_folder():
    '''Return user Desktop folder'''
    return shellapp.namespace(0).self.path
```
I put them in `windows.py` because they don't apply to any other OS. However logistically I see how they might fit better in `utils.py`. In `get_homedir()` The windows code for home would be: 

    shellapp.namespace(40).self.path   # 40 = user profile dir (`C:\Users\Matt` in Win10)

Note: there's nothing wrong with current cross-platform method of querying environment vars to find HOME or USERPROFILE. Where this changes is extrapolating from Userprofile to Desktop and Start Menu as they're not guaranteed to be underneath in the same place.

Which brings us to **shortcut.py** and murky waters. I did start down this road, but backed out because it was quickly turning into a "rewrite half the file" scenario and I got nervous about the scale of change. I'm okay with functions, but my brain hasn't grokked Classes yet. Can you provide some suggestions of next steps?

### CODE REVIEW

**Comment by maphew** [comment]
There's a better way yet of determining Windows special folders (better meaning more future proof), see https://github.com/maphew/pyshortcuts/tree/knownfolders. However it adds a whole new file and I don't think that's necessary for just Menu and Desktop determination. So this is just FYI.

**Comment by newville** [comment]
@maphew  I like this.   That is, I'm OK with keeping "Desktop shortcuts" and adding "Start Menu entries", especially on Windows.  I would be OK with an option to add Start Menu entries and *not* have Desktop icons, but could go either way.    That is, I think it's fine to have Desktop icons made when installing an app that a user can decide to throw away or move.

Also,  FWIW with #5 many (perhaps "most" -- hard to tell) of the options for Linux desktops that have something like a Start Menu will have entries added. 

The `knownfolders.py` code looks very useful.  If I read that and your comment on it correctly, one could probably simplify that to get just the path to Desktop and StartMenu and maybe CommonStartMenu.  I would definitely be OK with that.

You also raise a good point about the sort of mess that is `get_homedir()`.  It is probably better being separate for each platform.   
How about changing the code so that each platform has a `get_folders()` function that returned a tuple (perhaps named tuple?) with fully resolved paths for Home, Desktop, and StartMenu/ . Maybe something like:

```
folders = get_folders()
print("Home, Desktop, StartMenu ", folders.home, folders.desktop, folders.start_menu)
```
    

Anyway, I like the approach and would encourage you to keep going down this path.

**Comment by maphew** [comment]
Yes I'm okay with both Menu and Desktop as the standard method. It's easy to remove or sort desktop icons; much more clicky and involved for menus. Thanks for the encouragement! I'll explore platform specific get_folders() and see how far I get.

**Comment by maphew** [comment]
This is probably a good point to have a look and see if you're in agreement with my direction, especially regarding using **[black](https://black.readthedocs.io/en/stable/)** for formatting. 

I'm delaying integrating and trimming the knownfolders.py approach until after creating menu shortcuts is added.

**Comment by newville** [comment]
@maphew Thank! I like this.  And I have no complaints or preferences for single or double quotes. 

Would it make sense to move `util.get_win_folders()` to `windows.get_folders()`, and `utils.get_nix_folders()` to `linux.get_folders()` (and then have `darwin.py` import that from `linux.py`)?  I think that could put essentially all of the Windows-specific code into `windows.py` and similarly all the Unix-specific code in `linux.py`, with `darwin.py` being a variant?   I'm sure there could be other approaches too.   What's your take on that?

**Comment by maphew** [comment]
glad you're happy with it! I like the idea of keeping platform stuff together.

**Comment by maphew** [comment]
...but I haven't been able to get it to work. Pytest fails on importing the new functions. I don't understand the back and forth dot-filename imports. I've not used relative imports before and don't know how to work with them. I used the same syntax as what was there previously, I think.

**Comment by newville** [comment]
@maphew  I think the problem may be "circular imports".  That is, `windows.py` has
`from .utils import get_homedir` 

but `utils.py` has a conditional

`from  .windows import get_folders`

I would say just remove those platform-specific imports in `utils.py` and import them from the platform-specific file.

File: pyshortcuts/shortcut.py [hunk 1]
```diff
@@ -70,12 +70,20 @@ def __init__(self, script, name=None, description=None,
...
+        #     # if folder var has a value
+        #     #   and it matches the beginning of just created `desktop` var
+        #     #   reassign folder var to be substring of folder var
+        #     #   that is 1 char longer that desktop var
+        #     # reassign `dest` to be desktop path + folder path
+        #     if folder.startswith(desktop):
+        #         folder = folder[len(desktop)+1:]
+        #     dest = os.path.join(desktop, folder)
+
+        #ignore all above, just make dest same as folder:
```
Comment (maphew) [inline pyshortcuts/shortcut.py]: I had a really hard time understanding what challenge this piece of code was solving, and in the end it seems to be not necessary at all. But maybe it's only unnecessary on Windows?

**Comment by maphew** [comment]
The circular importing is fixed. The CI checks are failing because I forced an error in pytest in order to see the debug print() statements I sprinkled about. There's probably a smarter way to do that.

I had a really hard time understanding what challenge the now-commented out piece of code in shortcut.py L73:82 for 7e9a653 was solving, and in the end it seems to be not necessary at all. But maybe it's only unnecessary on Windows?

**Comment by maphew** [comment]
CI is happy with Windows and Linux. There's a bunch of deadwood to clear out but the approach seems to be working.

**Comment by newville** [comment]
@maphew  Great, thanks!  This looks like it is failing on MacOS, but I think I can fix that.  

FWIW, the `folder` option in `Shortcut.shortcut.py` is so that one can make a folder (for example "MyCoolApp") on the desktop and stick a bunch of shortcuts/links in that folder.   I actually use this, 
 as a couple of my projects produce multiple entry points / apps.   

I will happily merge this and expect to fix these small problems.  I might do a little bit of other re-arranging and API cleanups (so that `from pyshortcuts import get_homedir` works, for example).  I will also try to push out a new release within a couple of days.   

Thanks - this is really nice!

**Comment by maphew** [comment]
Ahhh, ok now I see. A grouping folder makes sense and I would use that too. I'll wait to see your refinements before I work on anything else.  Thanks for merging!

**Comment by newville** [comment]
@maphew Yeah, I went a little crazy ;).  I actually just started doing some tweaks and cleanups and then kept going with "ooh, what if we do this...".

**Pull request merged**